src/control/lib/spdk/src/nvme_control.c

```diff
@@ -65,121 +71,204 @@ get_dev_health_logs(struct spdk_nvme_ctrlr *ctrlr,
-			ns_id);
-		ret->rc = -NVMEC_ERR_NS_ID_UNEXPECTED;
-		return ret;
-	}
+static struct ret_t *
+wipe(char *ctrlr_pci_addr)
+{
+	struct ctrlr_entry	*centry;
+	struct ns_entry		*nentry;
+	struct ret_t		*ret;
+	struct wipe_sequence	 sequence;
+	int			 rc;
 
-	if (spdk_nvme_ctrlr_get_next_active_ns(sequence.ctrlr, ns_id) != 0) {
-		/* unexpected, don't wipe */
-		snprintf(ret->err, sizeof(ret->err),
-			"expected a single active namespace, got multiple");
-		ret->rc = -NVMEC_ERR_MULTIPLE_ACTIVE_NS;
-		return ret;
-	}
+	ret = init_ret(0);
 
-	sequence.ns = spdk_nvme_ctrlr_get_ns(sequence.ctrlr, ns_id);
-	if (!sequence.ns) {
-		snprintf(ret->err, sizeof(ret->err), "null ns");
-		ret->rc = -NVMEC_ERR_NULL_NS;
+	rc = get_controller(&centry, ctrlr_pci_addr);
+	if (rc != 0) {
+		snprintf(ret->info, sizeof(ret->info), "get_controller()\n");
+		ret->rc = rc;
 		return ret;
 	}
 
-	sequence.qpair = spdk_nvme_ctrlr_alloc_io_qpair(sequence.ctrlr, NULL,
-							0);
-	if (sequence.qpair == NULL) {
-		snprintf(ret->err, sizeof(ret->err),
-			"spdk_nvme_ctrlr_alloc_io_qpair() failed");
-		ret->rc = -NVMEC_ERR_ALLOC_IO_QPAIR;
-		return ret;
-	}
+	nentry = centry->nss;
+	while (nentry != NULL) {
+		nentry->qpair = spdk_nvme_ctrlr_alloc_io_qpair(nentry->ctrlr,
+							       NULL, 0);
+		if (nentry->qpair == NULL) {
+			snprintf(ret->info, sizeof(ret->info),
+				 "spdk_nvme_ctrlr_alloc_io_qpair()\n");
+			ret->rc = -1;
+			return ret;
+		}
 
-	sequence.using_cmb_io = 1;
-	sequence.buf = spdk_nvme_ctrlr_alloc_cmb_io_buffer(sequence.ctrlr,
-							   0x1000);
-	if (sequence.buf == NULL) {
-		sequence.using_cmb_io = 0;
 		sequence.buf = spdk_zmalloc(0x1000, 0x1000, NULL,
 					    SPDK_ENV_SOCKET_ID_ANY,
 					    SPDK_MALLOC_DMA);
+		if (sequence.buf == NULL) {
+			snprintf(ret->info, sizeof(ret->info),
+				 "spdk_zmalloc()\n");
+			ret->rc = -1;
+			spdk_nvme_ctrlr_free_io_qpair(nentry->qpair);
+			return ret;
+		}
+		sequence.is_completed = 0;
+		sequence.ns_entry = nentry;
+
+		rc = spdk_nvme_ns_cmd_write(nentry->ns, nentry->qpair,
+					    sequence.buf, 0, /* LBA start */
+					    1, /* number of LBAs */
+					    write_complete, &sequence, 0);
+		if (rc != 0) {
+			snprintf(ret->info, sizeof(ret->info),
+				 "spdk_nvme_ns_cmd_write() (%d)\n", rc);
+			ret->rc = -1;
+			spdk_free(sequence.buf);
+			spdk_nvme_ctrlr_free_io_qpair(nentry->qpair);
+			return ret;
+		}
+
+		while (!sequence.is_completed) {
+			rc = spdk_nvme_qpair_process_completions(nentry->qpair,
+								 0);
+			if (rc < 0) {
+				fprintf(stderr,
+					"process completions returns %d\n", rc);
+				break;
+			}
+		}
+
+		if (sequence.is_completed != 1) {
+			snprintf(ret->info, sizeof(ret->info),
+				 "spdk_nvme_ns_cmd_write() callback\n");
+			ret->rc = -1;
+			spdk_free(sequence.buf);
+			spdk_nvme_ctrlr_free_io_qpair(nentry->qpair);
+			return ret;
+		}
+
+		snprintf(ret->info + sizeof(ret->info),
+			 sizeof(TEXTLEN) - strlen(ret->info), "%d ",
+			 spdk_nvme_ns_get_id(nentry->ns));
```

Comment:
Is there a field after ret->info, or is this supposed to be writing in ret->info? sizeof(ret->info) != strlen(ret->info)

Could clarify a little more with nicely-named variables (even though they're not strictly necessary).

An aside, we are also supposed to be using strnlen instead of strlen, for safety reasons.

Reply:
this appends namespace ids to info to be printed go-side on success

Reply:
changed to use strnlen